### ISSUE TITLE
shell: make object construction less magic?

### ISSUE THREAD

**Issue by shykes** [dagger/dagger#8834]
## Problem

In `dagger call`, the user must express the value of each argument as single string value. In order to support object types (`Container`, `Directory`, `Service` etc), the CLI supports a "magic encoding" of these object types as a string. For example:

- A `Container` type can be encoded as a remote docker image address, for example `index.docker.io/alpine:latest`
- A `Directory` type can be encoded as either a remote git ref (`https://github.com/dagger/dagger#main:docs`) or a local directory (`./foo/bar`)
- etc.

Without this feature, `dagger call` would be simply useless. But, it is far from perfect. It doesn't support all possible values for objects; and the magic can be hard to explain and understand.

## Solution

With `dagger shell`, we have more options:

- *Pipelines as arguments* allow passing any object as argument. This works for core *and custom* objects
- *Builtins* allow exposing the full power of the core API as shell commands

In short: the dagger shell is expressive enough to no longer need magic encoding. So... maybe we should disable it?

| Before | After |
| -- | -- |
| `dagger call build --source ./frontend` | `build $(.context ./frontend)` |
| `dagger call build --source https://github.com/dagger/dagger` |  `build $(.git https://github.com/dagger/dagger)` |
| `dagger call test --db tcp://localhost:4242` | `test --db $(.tunnel tcp://localhost:4242)` |

etc

**Comment by shykes** [comment]
wdyt @helderco ?

**Comment by helderco** [comment]
I love the simplicity and explicitness 👍, I’m just concerned about the added verbosity. Both work interchangeably right now, so you can already avoid using the flags when working with `dagger shell` and showing it off in demos. Then see how it feels.
